src/vector_tile_load_tile.hpp

```diff
@@ -40,31 +63,50 @@ void merge_from_buffer(merc_tile & t, const char * data, std::size_t size)
             case Tile_Encoding::LAYERS:
                 {
                     auto layer_data = tile_msg.get_data();
-                    protozero::pbf_reader layer_valid_msg(layer_data);
-                    std::set<validity_error> errors;             
-                    std::string layer_name;
-                    std::uint32_t version = 1;
-                    layer_is_valid(layer_valid_msg, errors, layer_name, version);
-                    if (!errors.empty())
-                    {   
-                        std::string layer_errors;
-                        validity_error_to_string(errors, layer_errors);
-                        throw std::runtime_error(layer_errors);
+                    protozero::pbf_reader layer_props_msg(layer_data);
+                    auto layer_info = get_layer_name_and_version(layer_props_msg);
+                    std::string const& layer_name = layer_info.first;
+                    std::uint32_t version = layer_info.second;
+                    if (validate)
+                    {
+                        std::set<validity_error> errors;
+                        if (version < 1 || version > 2)
+                        {
+                            errors.insert(LAYER_HAS_UNSUPPORTED_VERSION);
+                        }
+                        if (t.has_layer(layer_name))
+                        {
+                            errors.insert(TILE_REPEATED_LAYER_NAMES);
+                        }
+                        protozero::pbf_reader layer_valid_msg(layer_data);
+                        layer_is_valid(layer_valid_msg, errors);
+                        if (!errors.empty())
+                        {
+                            std::string layer_errors;
+                            validity_error_to_string(errors, layer_errors);
+                            throw std::runtime_error(layer_errors);
+                        }
                     }
-                    if (t.has_layer(layer_name))
+                    else if (layer_name.empty() || t.has_layer(layer_name))
                     {
                         continue;
                     }
-                    if (version == 1)
+                    if (upgrade && version == 1)
+                    {
+                            // v1 tiles will be converted to v2
+                            protozero::pbf_reader layer_msg(layer_data);
+                            ds_vec.push_back(std::make_shared<mapnik::vector_tile_impl::tile_datasource_pbf>(
+                                        layer_msg,
+                                        t.x(),
+                                        t.y(),
+                                        t.z(),
+                                        true));
+                    }
+                    else if (validate && version != 1)
```

Comment:
@springmeyer did you mean `upgrade && version != 1`? If so, I think we should probably just quietly ignore the situation of `upgrade && version == 2`, but probably throw or warn when `upgrade && version == 3` or something along those lines comes up.